osmoutils/binary_search.go

```diff
@@ -0,0 +1,95 @@
+package osmoutils
+
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ErrTolerance is used to define a compare function, which checks if two
+// ints are within a certain error tolerance of one another.
+// ErrTolerance.Compare(a, b) returns true iff:
+// |a - b| <= AdditiveTolerance
+// |a - b| / min(a, b) <= MultiplicativeTolerance
+// Each check is respectively ignored if the entry is nil (sdk.Dec{}, sdk.Int{})
+type ErrTolerance struct {
+	AdditiveTolerance       sdk.Int
+	MultiplicativeTolerance sdk.Dec
+}
+
+// IsSatisfied returns if actual is within errTolerance of expected.
+// returns 0 if it is
+// returns 1 if not, and expected > actual.
+// returns -1 if not, and expected < actual
+func (e ErrTolerance) Compare(expected sdk.Int, actual sdk.Int) int8 {
```

Comment:
Some of the standard library `Compare` functions return `int` instead of `int8`. Functions taking compare function could expect type signature of return type `int`.
- https://pkg.go.dev/bytes#Compare
- https://pkg.go.dev/strings@go1.18.1#Compare
- https://pkg.go.dev/math/big@go1.18.1#Int.Cmp